elastic/elastic-package
# Implement "publish" command

Fixes: https://github.com/elastic/elastic-package/issues/21

This PR  implements "publish" command, which will replace "mage UpdatePackageStorage". It guarantees feature parity with https://github.com/elastic/integrations/tree/master/dev/update-package-storage

This improvement will introduce a better logging of performed actions and will allow for executing publish independently for every package (in parallel).

 

TODOs:

- [x] Read package context (especially manifest)
- [x] Check if the package has been built
- [x] Refactor "promote" to reuse as much as possible
- [x] Checkout package-storage
- [x] Find greatest package revision released (revision + stage)
- [x] Copy greatest package to staging as new revision
- [x] Branch out from snapshot
- [x] Commit the above
- [x] Copy over the built package
- [x] Commit the above
- [x] Push to remote Git branch
- [x] Open PR
- [x] Fix: check if package is already released

Sample PR:
https://github.com/elastic/package-storage/pull/1054

Testing:

Unfortunately this time we need to perform some testing:
1. Publish new package to the package storage (select any from `test/packages`).
2. Publish updated package to the package storage (e.g. `nginx-999.999.999`).
3. Try to publish package while the package has been already published (a candidate in integrations?).
4. Try to publish package while the PR is open (e.g. `nginx-999.999.999`).
5. Promote packages between stages, as refactoring impacts that feature. 

Sample output:

```
elastic-package publish -v
elastic-package has been installed.
2021/03/23 14:09:37 DEBUG Enable verbose logging
Publish the package
Current GitHub user: mtojek
Use build directory: /Users/marcin.tojek/go/src/github.com/elastic/elastic-package/build/integrations/nginx/999.999.999
Clone package-storage repository
Find latest package revision of "nginx" in package-storage
2021/03/23 14:09:55 DEBUG Change stage to production
2021/03/23 14:10:04 DEBUG Find revisions of the "nginx" package
2021/03/23 14:10:04 DEBUG Found package revision: nginx-0.2.3
2021/03/23 14:10:04 DEBUG Found package revision: nginx-0.2.4
2021/03/23 14:10:04 DEBUG Found package revision: nginx-0.3.6
2021/03/23 14:10:04 DEBUG Found package revision: nginx-0.3.7
2021/03/23 14:10:04 DEBUG Found package revision: nginx-0.3.9
2021/03/23 14:10:04 DEBUG Change stage to staging
2021/03/23 14:10:13 DEBUG Find revisions of the "nginx" package
2021/03/23 14:10:13 DEBUG Found package revision: nginx-0.4.0
2021/03/23 14:10:13 DEBUG Change stage to snapshot
2021/03/23 14:10:17 DEBUG Find revisions of the "nginx" package
Latest package revision: nginx-0.4.0 (stage: staging)
Copy sources of the latest package revision to index
2021/03/23 14:10:17 DEBUG Checkout source stage: staging
2021/03/23 14:10:20 DEBUG Load package resources from source stage
2021/03/23 14:10:20 DEBUG Checkout destination stage: snapshot
2021/03/23 14:10:23 DEBUG Create new destination branch: update-snapshot-1616505017944000000
2021/03/23 14:10:25 DEBUG Write package resources to destination branch
2021/03/23 14:10:25 DEBUG Add package resources to index
2021/03/23 14:10:32 DEBUG Commit changes to destination branch
2021/03/23 14:10:33 DEBUG Temporarily remove all files from index
2021/03/23 14:10:33 DEBUG Evaluate all resource paths for the package (buildDir: /Users/marcin.tojek/go/src/github.com/elastic/elastic-package/build/integrations)
2021/03/23 14:10:33 DEBUG Add updated resources to index
2021/03/23 14:10:40 DEBUG Commit changes to destination branch
Push new package revision to storage
Check if pull request is already open
2021/03/23 14:10:56 DEBUG Use Search API to find an open pull request (query: 'repo:elastic/package-storage base:snapshot is:pr is:open in:title "[snapshot] Update "nginx" integration to version 999.999.999"')
Open new pull request
2021/03/23 14:10:57 DEBUG Create new pull request (head: mtojek:update-snapshot-1616505017944000000, base: snapshot)
2021/03/23 14:10:58 DEBUG Pull request URL: https://api.github.com/repos/elastic/package-storage/pulls/1054
2021/03/23 14:10:58 DEBUG Update reviewers (pull request ID: 1054)
2021/03/23 14:10:58 DEBUG Reviewers requested successfully (teams: 1, reviewers: 0)
Done
```
